### PR TITLE
Convert CMake to use git submodules instead of FetchContent

### PR DESCRIPTION
**Pull request by maxrossi91**
What was the issue in using CMake over submodules? Is this because you want to control the dependencies directly into the CMake?

@rvarki please open pull requests to all the submodules you updated, thanks.

### CODE REVIEW

File: CMakeLists.txt [hunk 1]
```diff
@@ -1,41 +1,323 @@
...
+list(APPEND CMAKE_MODULE_PATH "${CMAKE_CURRENT_BINARY_DIR}") #find_package
+list(APPEND CMAKE_PREFIX_PATH "${CMAKE_LIBRARY_OUTPUT_DIRECTORY}" "${CMAKE_ARCHIVE_OUTPUT_DIRECTORY}") #find_library
 
-# Configure thirdparty
-# ------------------------------------------------------------------------------
-set(CMAKE_INSTALL_INCLUDEDIR "include") # This is an hack because include(GUIInstallDirs) doesn't work
+# ZLIB
+find_package(ZLIB REQUIRED)
+
+# SDSL
```
Comment (maxrossi91) [inline CMakeLists.txt:62]: Please move the thirdparty related components into a CMakeList.txt in the `thirdparty` folder so it is cleaner to read.

**Review by maxrossi91 — COMMENTED** [review]
This is not the most elegant solution, but as long as it works, it is fine. I think the `FetchContent` approach is cleaner and delegate responsibilities on the single components, other then re-building the whole compile structure.

**Comment by rvarki** [comment]
The idea to switch from FetchContent to git submodules came from Pjotr in our discussion in Element with Erik Garrison's group about how to get it to get moni-align compile on other machines. He was really pushing the idea to make the switch so I was trying it out on this branch. There are pros and cons to both approaches. With FetchContent, like you said you can build the projects in isolation however if the projects build the same targets then you spend rebuilding the same targets again. With git submodules the targets are exposed globally which requires dealing with target conflicts from the different projects but with the upside of faster compiling and better visibility if dealt with correctly. 

I agree this branch is not necessarily as elegant as the FetchContent from my perspective as well. There now exists an implicit level of dependency between the projects in this branch since there are shared target dependencies. I should probably expose the target dependencies in the main cmake. It's why running make works but not make -j when compiling this branch. I also make the assumption that targets built with the same name in different subprojects are really the same targets and therefore only build the targets that haven't been built yet in each subproject which could be wrong assumption. 

I have another branch called update_build which just directly upgrades the dependencies of a few subprojects in the thirdparty cmake and I can compile with newer gcc versions on hipergator but Omar had issues compiling on their system with that branch. However, this branch has made me aware of a few potential issues with the current cmake in the main branch which I could also try to incorporate in the update_build branch and see if it makes it compatible if we wanted to upgrade while still using FetchContent philosophy in the main project. I really don't know which best way to do it is. 

Let's hold on merging for now. Andrea still has some issues running the compiled version of this branch on their cluster, but it works on his laptop. In the meantime I can also try to get the update_build branch to work and the we can choose which way we want to upgrade (stick with FetchContent or switch to git submodules).

**Comment by maxrossi91** [comment]
@AndreaGuarracino let's wait on @rvarki to give the ok for this PR to be merged.

**Comment by rvarki** [comment]
@maxrossi91 I am okay with this getting merged. I checked with Omar and he is also able to build the project on the JHU servers as well from this branch. However, we should merge into develop and then into master like we have been doing with the squash command.

**Pull request merged**